docs/internal/concurrency/modelimpl/StableOrder.cs

```diff
@@ -0,0 +1,49 @@
+/**
+ *  Used to provide a stable order over object.
+ * 
+ *  It initially uses HashCode to provide an order. If there 
+ *  is a hashcode collision, then we use a counter to disambiguate between
+ *  the two objects.
+ * 
+ *  Alternatively, we could just use an atomic on allocation. The design here
+ *  aims to reduce the likelyhood of having to perform the atomic increment. 
+ **/
+class StableOrder : IComparable
+{
+    // Used to disambiguate between two cowns that have the same hashcode.
+    private static volatile int counter = 0;
+
+    // If we get a hashcode collision, then we need to use an additional value
+    // to distinguish between the two objects. We don't want to use this from
+    // the start as it requires Interlocked operations to create it.
+    private Int64 collision = 0;
+
+    private void EnsureCollisionSet()
+    {
+        if (collision == 0)
```

Comment:
Doesn't this need a CAS loop over `collision` or similar? What if two threads are scheduling a behaviour over the same cown concurrently?

Reply:
Good spot.  Thanks. I'll fix that.

Reply:
> Alternatively, we could just use an atomic on allocation

FWIW, IMO if the goal is explaining the semantics, then this is the simpler and thus right route to take. No need to worry about collisions f.ex.

Reply:
Okay. I'll update this class to be less fun ;-)

Reply:
Sorry! ;)